### ISSUE TITLE
Subtitles disappear after seeking backwards and toggling subs when caching is enabled

### ISSUE THREAD

**Issue by pew-pew** [mpv-player/mpv#9646]
### Important Information

Provide following Information:
- mpv version
```
$ mpv --version
mpv 0.34.0 Copyright © 2000-2021 mpv/MPlayer/mplayer2 projects
 built on UNKNOWN
FFmpeg library versions:
   libavutil       56.70.100
   libavcodec      58.134.100
   libavformat     58.76.100
   libswscale      5.9.100
   libavfilter     7.110.100
   libswresample   3.9.100
FFmpeg version: n4.4.1
```
- Linux Distribution and Version
    - manjaro linux 21.2.0, kernel 5.10.79-1-MANJARO, x86-64
- Source of the mpv binary
    - arch package `mpv` `1:0.34.0-2`
- If known which version of mpv introduced the problem
    - don't know
- Window Manager and version
    - xfce 4.16
- GPU driver and version
    - probably video-hybrid-intel-nvidia-470xx-prime 2021.11.04
- Possible screenshot or video of visual glitches
    - not relevant

### Reproduction steps

0. Obtain sample video file:
    - run this python (>=3.6) script https://0x0.st/orj9.py
    - this script will create `video.mkv` file (**600MB** in size) - it has one audio and one subtitle stream
1. run `mpv --force-window --msg-level=mkv=trace --demuxer-max-bytes=10MiB --cache=yes video.mkv`
2. Seek somewhere in the middle (30:00)
3. Wait (probably less than one second) until log shows that demuxer buffer is filled (`[mkv] bytes = 10xxxxxx ...`)
4. Seek backward for 10 seconds or more (e.g. by pressing `left arrow` a few times)
5. Ensure that subtitles are visible (something like `00:28:22,000 to 00:28:22,490`)
6. Ensure that log shows overflow of demuxer buffer (`[mkv] bytes = X ...` where X is larger than 10MiB)
7. Toggle subtitles twice (press `j` twice) - this should switch to `Subtitles: no` and then back to `Subtitles: (1) unknown`

### Expected behavior

After step 7 subtitles are displayed.

### Actual behavior

After step 7 subtitles are not displayed.

### Log file

https://0x0.st/orlD.txt

### Sample files

Not relevant, use the script referenced in `Reproduction steps` to generate one.

### Note

I regularly face this issue during playback of large (>1GB) movies from my slow HDD mounted with ntfs-3g. I frequently switch between subtitles, but sometimes this action makes subtitles disappear completely until some random point in future or a large seek, and it forces me to restart mpv.

**Comment by LaserEyess** [comment]
Can you try [`--demuxer-mkv-subtitle-preroll`](https://mpv.io/manual/master/#options-demuxer-mkv-subtitle-preroll)?

**Comment by pew-pew** [comment]
> Can you try [`--demuxer-mkv-subtitle-preroll`](https://mpv.io/manual/master/#options-demuxer-mkv-subtitle-preroll)?

Hi, this option doesn't help

**Comment by pew-pew** [comment]
Just tried with .mp4 container and embedded subtitles (mov_text "codec") - can reproduce the issue:
```ffmpeg -f lavfi -t 3600 -i anullsrc=channel_layout=stereo:sample_rate=44100 -i ./subtitles.srt -c:s mov_text -c:a ac3 -b:a 600k ./video.mp4```
(it's 250mb now, and input file can also be .ass)

**Comment by LaserEyess** [comment]
I can reproduce this on an artificially slow video source, in this case ssh to a machine with 100 ms of latency. Maybe this is some sort of race condition with backwards seeking with respect to subtitles.

**Comment by pew-pew** [comment]
With the help of git-bisect I have found that since commit c59ca06a0fff432ac4cae012bb0299a8db9a00d3 (between v0.32 and v0.33) this issue can be reproduced with very high probability. Before that commit this behavior still occurs, but the odds are very low (earliest version I have checked is v0.30.0).

**Comment by pew-pew** [comment]
It seems that side-effect of seeking to the end of the file and back in `stream/stream_file.c:get_size` was somehow important - adding two following lines make the issue disappear:
```diff
static int64_t get_size(stream_t *s)
{
    struct priv *p = s->priv;
    off_t size = lseek(p->fd, 0, SEEK_END);
    lseek(p->fd, s->pos, SEEK_SET);
    return size == (off_t)-1 ? -1 : size;
    if (p->cached_size == -2) {
        off_t size = lseek(p->fd, 0, SEEK_END);
        lseek(p->fd, s->pos, SEEK_SET);
        p->cached_size = size < 0 ? -1 : size;
    }
+    lseek(p->fd, 0, SEEK_END);
+    lseek(p->fd, s->pos, SEEK_SET);
    return p->cached_size;
}
```
(`p->cached_size` never equals -1 in my case)

Hovewer, adding only the second line (i.e. only seeking to `s->pos`, but not to the end) isn't sufficient. I have checked that at the time of calling of this function current offset inside the file is the same as `s->pos`. At the moment I don't know what important side effect the action of seeking to the end of the file has.

**Comment by pew-pew** [comment]
I've tried to debug (on master), and have found following:

I have monitored demuxer's queue size (`get_foward_buffered_bytes`, `ds->reader_head`, `ds->queue->tail`).

During usual playback, total forwarded bytes is approximately demuxer's max_bytes field (idk what is this, it's 150mb in my case), and the distance between reader head and queue tail is kept so that we don't exceed the max_size.

But sometimes if I aggressively seek backwards (i.e. taking steps from the issue description), total forwarded bytes is way larger than max_size, queue tail is not moving anywhere, reader head is slowly moving with the speed of video playback. And If continue seeking backwards (few times), reader head is moved backward, but queue tail doesn't move at all, so queue size grows.

Inside [`read_packet`](https://github.com/mpv-player/mpv/blob/master/demux/demux.c#L2178) function (`demux/demux.c`) variable `read_more` is false when this happens, and I am not sure if this ok or not (since [this code](https://github.com/mpv-player/mpv/blob/2157bb1995334e25bd865b438e23764ff966cacb/demux/demux.c#L2217-L2239) looks like I should be warned about too many packets in the queue.

**Comment by pew-pew** [comment]
I was able to reproduce the issue on any storage by providing explicit `--cache=yes` flag. I have updated reproduction steps and issue title.

**Comment by pew-pew** [comment]
Setting `--cache=no` in the command line made the issue almost disappear - during ~2 hours of playback with frequent small seeks and subtitle toggles, I have struck in a situation with lost subtitles only once. I have made core dump in this state just in case.

**Comment by pew-pew** [comment]
I've tried to debug this and come to the following explanation of the issue reason:
- To show subtitles after subtitles switch, mpv has to read them from the file.
- mpv allows reading from the file [only if](https://github.com/mpv-player/mpv/blob/2157bb1995334e25bd865b438e23764ff966cacb/demux/demux.c#L2215) queue (buffered data of all streams) is not overflowed
- But seeking backwards (when caching is enabled) easily triggers queue overflow because mpv doesn't shrink cache window after backwards seek, it just prepends new data (at least until certain threshold)
- By the way, if you try to switch subs during regular playback, it [will](https://github.com/mpv-player/mpv/blob/2157bb1995334e25bd865b438e23764ff966cacb/demux/demux.c#L3973) trigger empty of all accumulated data, but it doesn't happen if queue is overflowed (I don't understand logic behind [`demux_internal->after_seek`](https://github.com/mpv-player/mpv/blob/2157bb1995334e25bd865b438e23764ff966cacb/demux/demux.c#L229) field).

I think that the main problem is that mpv doesn't truncate cache window after seeking backwards, but I'm probably wrong since I don't know anything about internals of mpv.

**Comment by Dudemanguy** [comment]
I'll admit I didn't go through these exact reproduction steps, but I fixed a bunch of related "subtitles disappear while toggling" issues a while ago so I'm going to assume this works now. If not, we can reopen.

**Comment by awsms** [comment]
mpv sometimes drop a line of the subtitles, and the only way to "restore" that missing line is to switch to another subtrack, and re-enable the previous one.
Been having this issue for years, but it happens so rarely I never bothered reporting, and I've never been able to find a way to reproduce it.
AFAIK this only happened when playing media on a sshfs mount (caching enabled).

latest repro:
- mpv was paused at 1040.122s
- the missing cue 278 should cover 1039.747 -> 1041.666.
- but IPC returned:
      - sub-text: empty
      - sub-text/ass: empty
      - sub-text/ass-full: empty
      - sub-start / sub-end: unavailable
      - sub-lines: missing that cue entirely

- sub-lines jumped from:
      - 1024.732 -> 1026.025 "Hmm?"
      - directly to 1041.666 -> 1043.334 "Why isn't anyone picking up?"

at that moment mpv's subtitle state itself did not contain line 278, despite the selected track being the normal English subrip stream.

the file played is an mkv, the srt is muxed-in. the srt where the issue just occurred:
```
  277
  00:17:04,732 --> 00:17:06,025
  Hmm?

  278
  00:17:19,747 --> 00:17:21,666
  This is Design Team 1 of Nana Atelier--

  279
  00:17:21,666 --> 00:17:23,334
  Why isn't anyone picking up?
```

**Comment by Dudemanguy** [comment]
Could you open a new issue for this specific one?
